File: vendingmachine.cpp
#include <iostream>
#include <vector>
#include <string>
#include <thread>
#include <chrono>
#include <sstream>
#include <iomanip> // Pour l'alignement du texte
#include <limits>  // Pour vider le buffer cin

// =============================================================
// 1. OUTILS VISUELS (UI & COULEURS)
// =============================================================

namespace Color {
    const std::string RESET   = "\033[0m";
    const std::string RED     = "\033[1;31m";
    const std::string GREEN   = "\033[1;32m";
    const std::string YELLOW  = "\033[1;33m";
    const std::string BLUE    = "\033[1;34m";
    const std::string CYAN    = "\033[1;36m";
    const std::string MAGENTA = "\033[1;35m";
    const std::string GRAY    = "\033[1;90m";
}

void clearScreen() {
    // Séquence ANSI pour effacer l'écran (fonctionne sur Linux/Mac et Win10+)
    std::cout << "\033[2J\033[1;1H";
}

void drawHeader() {
    std::cout << Color::GREEN;
    std::cout << "╔═══════════════════════════════════════════════╗\n";
    std::cout << "║         DISTRIBUTEUR AUTOMATIQUE 3000         ║\n";
    std::cout << "╚═══════════════════════════════════════════════╝\n";
    std::cout << Color::RESET;
}

// =============================================================
// 2. TYPES DE DONNÉES (DATA)
// =============================================================

// Une Enum pour les types de produits (utile pour le portage Rust !)
enum class Category { DRINK, SNACK, TECH };

struct Product {
    int id;
    std::string name;
    float price;
    int stock;
    Category type;

    // Helper pour afficher l'icône selon la catégorie
    std::string getIcon() const {
        switch(type) {
            case Category::DRINK: return "[Boisson]";
            case Category::SNACK: return "[ Snack ]";
            case Category::TECH:  return "[ Tech  ]";
            default: return "[ ? ]";
        }
    }
};

// =============================================================
// 3. ABSTRACTION MATÉRIELLE (DRIVERS)
// =============================================================

class Hardware {
public:
    // Simulation d'un écran LCD 2 lignes
    static void lcdPrint(const std::string& line1, const std::string& line2 = "") {
        std::cout << "\n" << Color::BLUE << "  [LCD] ┌───────────────────────────────────────┐" << Color::RESET << "\n";
        std::cout << Color::BLUE << "  [LCD] │ " << Color::YELLOW << std::left << std::setw(37) << line1 << Color::BLUE << " │" << Color::RESET << "\n";
        if (!line2.empty()) {
             std::cout << Color::BLUE << "  [LCD] │ " << Color::YELLOW << std::left << std::setw(37) << line2 << Color::BLUE << " │" << Color::RESET << "\n";
        }
        std::cout << Color::BLUE << "  [LCD] └───────────────────────────────────────┘" << Color::RESET << "\n\n";
    }

    // Simulation du moteur avec barre de progression
    static void activateMotor(const std::string& productName) {
        std::cout << Color::MAGENTA << "  [MOTEUR] Livraison en cours : " << productName << Color::RESET << "\n  ";
        
        for (int i = 0; i <= 20; ++i) {
            std::cout << "▓";
            std::cout.flush();
            std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Vitesse moteur
        }
        std::cout << " OK!\n";
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
        
        std::cout << Color::GREEN << "  >>> CLONG ! Le produit est tombé dans le bac. <<<" << Color::RESET << "\n";
        std::this_thread::sleep_for(std::chrono::milliseconds(1500));
    }

    // Simulation du monnayeur (Accepteur de pièces)
    static void coinSound() {
        std::cout << Color::GRAY << "  (Clink! Pièce acceptée...)" << Color::RESET << "\n";
        std::this_thread::sleep_for(std::chrono::milliseconds(300));
    }
};

// =============================================================
// 4. LOGIQUE MÉTIER (CORE)
// =============================================================

class VendingMachine {
private:
    std::vector<Product> inventory;
    float currentCredit = 0.0f;

public:
    VendingMachine() {
        // Initialisation de l'inventaire complet
        inventory = {
            {10, "Coca-Cola Zero", 1.20f, 5, Category::DRINK},
            {11, "Ice Tea Peche ", 1.40f, 4, Category::DRINK},
            {12, "Eau Minerale  ", 0.80f, 8, Category::DRINK},
            {20, "Kinder Bueno  ", 1.10f, 6, Category::SNACK},
            {21, "M&Ms Peanut   ", 1.30f, 3, Category::SNACK},
            {22, "Chips Nature  ", 1.00f, 2, Category::SNACK},
            {30, "Cable USB-C   ", 5.50f, 2, Category::TECH},
            {31, "Ecouteurs     ", 8.00f, 1, Category::TECH}
        };
    }

    void run() {
        while (true) {
            clearScreen();
            drawHeader();
            displayInventory();
            displayControls();
            
            // Affichage du crédit actuel sur le LCD
            std::string creditMsg = "CREDIT : " + formatMoney(currentCredit);
            Hardware::lcdPrint("Veuillez choisir un produit", creditMsg);

            handleUserAction();
        }
    }

private:
    std::string formatMoney(float amount) {
        std::stringstream ss;
        ss << std::fixed << std::setprecision(2) << amount << " EUR";
        return ss.str();
    }

    void displayInventory() {
        std::cout << "  ID  | TYPE      | NOM             | PRIX    | STOCK \n";
        std::cout << "  ----+-----------+-----------------+---------+-------\n";
        
        for (const auto& item : inventory) {
            // Choix de la couleur selon le stock
            std::string stockColor = (item.stock > 0) ? Color::GREEN : Color::RED;
            std::string stockText  = (item.stock > 0) ? std::to_string(item.stock) : "RUPTURE";

            std::cout << "  " << Color::CYAN << std::setw(3) << item.id << Color::RESET << " | "
                      << item.getIcon() << " | "
                      << std::left << std::setw(15) << item.name << " | "
                      << Color::YELLOW << std::fixed << std::setprecision(2) << item.price << " €" << Color::RESET << " | "
                      << stockColor << stockText << Color::RESET << "\n";
        }
        std::cout << "\n";
    }

    void displayControls() {
        std::cout << "  [PIECES] : 1 = 0.10 EUR, 2 = 0.20 EUR, 3 = 0.50 EUR, 4 = 1.00 EUR, 5 = 2.00 EUR\n";
        std::cout << "  [ACTION] : Tapez l'ID du produit pour acheter\n";
        std::cout << "  [SYSTEM] : 99 pour Rendre monnaie / 0 pour Quitter\n";
        std::cout << Color::GRAY << "  Votre choix > " << Color::RESET;
    }

    void handleUserAction() {
        int input;
        std::cin >> input;

        // Gestion des erreurs de saisie (si l'utilisateur tape des lettres)
        if (std::cin.fail()) {
            std::cin.clear();
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return;
        }

        // --- GESTION DES PIÈCES ---
        if (input >= 1 && input <= 5) {
            float coinValue = 0.0f;
            switch(input) {
                case 1: coinValue = 0.10f; break;
                case 2: coinValue = 0.20f; break;
                case 3: coinValue = 0.50f; break;
                case 4: coinValue = 1.00f; break;
                case 5: coinValue = 2.00f; break;
            }
            Hardware::coinSound();
            currentCredit += coinValue;
            return;
        }

        // --- GESTION SYSTÈME ---
        if (input == 0) exit(0);
        
        if (input == 99) {
            if (currentCredit > 0) {
                Hardware::lcdPrint("RENDU MONNAIE...", formatMoney(currentCredit));
                std::this_thread::sleep_for(std::chrono::milliseconds(2000));
                currentCredit = 0.0f;
            }
            return;
        }

        // --- GESTION ACHAT ---
        processPurchase(input);
    }

    void processPurchase(int productId) {
        // Recherche du produit
        Product* selected = nullptr;
        for (auto& item : inventory) {
            if (item.id == productId) {
                selected = &item;
                break;
            }
        }

        if (selected == nullptr) {
            Hardware::lcdPrint("ERREUR", "ID Inconnu !");
            std::this_thread::sleep_for(std::chrono::milliseconds(1500));
            return;
        }

        if (selected->stock <= 0) {
            Hardware::lcdPrint("RUPTURE DE STOCK", "Choisissez un autre");
            std::this_thread::sleep_for(std::chrono::milliseconds(1500));
            return;
        }

        if (currentCredit < selected->price) {
            float missing = selected->price - currentCredit;
            Hardware::lcdPrint("CREDIT INSUFFISANT", "Manque : " + formatMoney(missing));
            std::this_thread::sleep_for(std::chrono::milliseconds(2000));
            return;
        }

        // Tout est OK : Transaction
        Hardware::lcdPrint("DISTRIBUTION...", selected->name);
        
        // Logique Transactionnelle
        currentCredit -= selected->price;
        selected->stock--;

        // Activation Hardware
        Hardware::activateMotor(selected->name);
        
        // Rendu monnaie automatique si on veut (optionnel)
        // Ici on garde le crédit pour un autre achat
    }
};

// =============================================================
// 5. POINT D'ENTRÉE
// =============================================================

int main() {
    // Config console pour afficher les caractères spéciaux si besoin (sur Windows vieux)
    // system("chcp 65001"); 
    
    VendingMachine machine;
    machine.run();
    return 0;
}

File: vendingmachine.rs
use std::io::{self, Write};
use std::{thread, time};

// =============================================================
// 1. OUTILS VISUELS (UI & COULEURS)
// =============================================================
mod color {
    pub const RESET: &str = "\x1b[0m";
    pub const RED: &str = "\x1b[1;31m";
    pub const GREEN: &str = "\x1b[1;32m";
    pub const YELLOW: &str = "\x1b[1;33m";
    pub const BLUE: &str = "\x1b[1;34m";
    pub const MAGENTA: &str = "\x1b[1;35m";
    pub const CYAN: &str = "\x1b[1;36m";
    pub const GRAY: &str = "\x1b[1;90m";
}

fn clear_screen() {
    print!("\x1b[2J\x1b[1;1H");
}

fn draw_header() {
    println!("{}{}", color::GREEN, "╔═══════════════════════════════════════════════╗");
    println!("║         DISTRIBUTEUR AUTOMATIQUE 3000         ║");
    println!("╚═══════════════════════════════════════════════╝{}", color::RESET);
}

// =============================================================
// 2. TYPES DE DONNÉES (DATA)
// =============================================================

#[derive(Debug, Clone, PartialEq)]
enum Category {
    Drink,
    Snack,
    Tech,
}

#[derive(Clone)]
struct Product {
    id: i32,
    name: String,
    price: f32,
    stock: i32,
    category: Category,
}

impl Product {
    fn get_icon(&self) -> &str {
        match self.category {
            Category::Drink => "[Boisson]",
            Category::Snack => "[ Snack ]",
            Category::Tech => "[ Tech  ]",
        }
    }
}

// =============================================================
// 3. ABSTRACTION MATÉRIELLE (DRIVERS)
// =============================================================

struct Hardware;

impl Hardware {
    fn lcd_print(line1: &str, line2: &str) {
        println!("\n{}  [LCD] ┌───────────────────────────────────────┐{}", color::BLUE, color::RESET);
        println!("{}  [LCD] │ {}{: <37}{} │{}", color::BLUE, color::YELLOW, line1, color::BLUE, color::RESET);
        if !line2.is_empty() {
            println!("{}  [LCD] │ {}{: <37}{} │{}", color::BLUE, color::YELLOW, line2, color::BLUE, color::RESET);
        }
        println!("{}  [LCD] └───────────────────────────────────────┘{}\n", color::BLUE, color::RESET);
    }

    fn activate_motor(product_name: &str) {
        println!("{}  [MOTEUR] Livraison en cours : {}{}", color::MAGENTA, product_name, color::RESET);
        print!("  ");
        io::stdout().flush().unwrap();

        for _ in 0..20 {
            print!("▓");
            io::stdout().flush().unwrap();
            thread::sleep(time::Duration::from_millis(100));
        }
        println!(" OK!");
        thread::sleep(time::Duration::from_millis(500));
        println!("{}  >>> CLONG ! Le produit est tombé dans le bac. <<<{}", color::GREEN, color::RESET);
        thread::sleep(time::Duration::from_millis(1500));
    }

    fn coin_sound() {
        println!("{}  (Clink! Pièce acceptée...){}", color::GRAY, color::RESET);
        thread::sleep(time::Duration::from_millis(300));
    }
}

// =============================================================
// 4. LOGIQUE MÉTIER (CORE)
// =============================================================

struct VendingMachine {
    inventory: Vec<Product>,
    current_credit: f32,
}

impl VendingMachine {
    fn new() -> VendingMachine {
        VendingMachine {
            current_credit: 0.0,
            inventory: vec![
                Product { id: 10, name: String::from("Coca-Cola Zero"), price: 1.20, stock: 5, category: Category::Drink },
                Product { id: 11, name: String::from("Ice Tea Peche "), price: 1.40, stock: 4, category: Category::Drink },
                Product { id: 12, name: String::from("Eau Minerale  "), price: 0.80, stock: 8, category: Category::Drink },
                Product { id: 20, name: String::from("Kinder Bueno  "), price: 1.10, stock: 6, category: Category::Snack },
                Product { id: 21, name: String::from("M&Ms Peanut   "), price: 1.30, stock: 3, category: Category::Snack },
                Product { id: 22, name: String::from("Chips Nature  "), price: 1.00, stock: 2, category: Category::Snack },
                Product { id: 30, name: String::from("Cable USB-C   "), price: 5.50, stock: 2, category: Category::Tech },
                Product { id: 31, name: String::from("Ecouteurs     "), price: 8.00, stock: 1, category: Category::Tech },
            ],
        }
    }

    fn run(&mut self) {
        loop {
            clear_screen();
            draw_header();
            self.display_inventory();
            self.display_controls();

            let credit_msg = format!("CREDIT : {:.2} EUR", self.current_credit);
            Hardware::lcd_print("Veuillez choisir un produit", &credit_msg);

            self.handle_user_action();
        }
    }

    fn display_inventory(&self) {
        println!("  ID  | TYPE      | NOM               | PRIX    | STOCK ");
        println!("  ----+-----------+-------------------+---------+-------");

        for item in &self.inventory {
            let (stock_color, stock_text) = if item.stock > 0 {
                (color::GREEN, item.stock.to_string())
            } else {
                (color::RED, "RUPTURE".to_string())
            };

            println!("  {}{:3}{} | {} | {: <15} | {}{:.2} €{} | {}{}{}",
                color::CYAN, item.id, color::RESET,
                item.get_icon(),
                item.name,
                color::YELLOW, item.price, color::RESET,
                stock_color, stock_text, color::RESET
            );
        }
        println!();
    }

    fn display_controls(&self) {
        println!("  [PIECES] : 1 = 0.10, 2 = 0.20, 3 = 0.50, 4 = 1.00, 5 = 2.00");
        println!("  [ACTION] : Tapez l'ID du produit pour acheter");
        println!("  [SYSTEM] : 99 pour Rendre monnaie / 0 pour Quitter");
        print!("{}  Votre choix > {}", color::GRAY, color::RESET);
        io::stdout().flush().unwrap();
    }

    fn handle_user_action(&mut self) {
        let mut input_str = String::new();
        
        // Lecture sécurisée
        if io::stdin().read_line(&mut input_str).is_err() {
            return;
        }

        // Parsing sécurisé avec gestion d'erreur (Result)
        let input: i32 = match input_str.trim().parse() {
            Ok(num) => num,
            Err(_) => return, // Si ce n'est pas un nombre, on ignore
        };

        match input {
            0 => std::process::exit(0),
            1..=5 => {
                let coin_value = match input {
                    1 => 0.10, 2 => 0.20, 3 => 0.50, 4 => 1.00, 5 => 2.00, _ => 0.0
                };
                Hardware::coin_sound();
                self.current_credit += coin_value;
            },
            99 => {
                if self.current_credit > 0.0 {
                    Hardware::lcd_print("RENDU MONNAIE...", &format!("{:.2} EUR", self.current_credit));
                    thread::sleep(time::Duration::from_millis(2000));
                    self.current_credit = 0.0;
                }
            },
            _ => self.process_purchase(input),
        }
    }

    fn process_purchase(&mut self, product_id: i32) {
        let index_opt = self.inventory.iter().position(|p| p.id == product_id);

        if let Some(index) = index_opt {
            let item = &mut self.inventory[index];

            if item.stock <= 0 {
                Hardware::lcd_print("RUPTURE DE STOCK", "Choisissez un autre");
                thread::sleep(time::Duration::from_millis(1500));
                return;
            }

            if self.current_credit < item.price {
                let missing = item.price - self.current_credit;
                Hardware::lcd_print("CREDIT INSUFFISANT", &format!("Manque : {:.2} EUR", missing));
                thread::sleep(time::Duration::from_millis(2000));
                return;
            }

            // Transaction et Clone pour éviter les soucis de Borrowing
            let product_name = item.name.clone(); 
            
            Hardware::lcd_print("DISTRIBUTION...", &product_name);
            
            item.stock -= 1;
            self.current_credit -= item.price;

            Hardware::activate_motor(&product_name);

        } else {
            Hardware::lcd_print("ERREUR", "ID Inconnu !");
            thread::sleep(time::Duration::from_millis(1500));
        }
    }
}

fn main() {
    let mut machine = VendingMachine::new();
    machine.run();
}
